File: MorySnip/Classes/Layers/Layer_Free.vb
Public Class Layer_Free
    Inherits Layer
    Public ReadOnly Property Path As New Drawing2D.GraphicsPath

    Public Sub New()

    End Sub

    Public Sub New(Pen As Pen, Brush As Brush, FirstPoint As Point)
        Me.Pen = Pen
        Me.Brush = Brush
        Me.Start(FirstPoint)
    End Sub

    Public Overrides Sub [Step](StepPoint As Point)
        Me.Path.AddLine(Me.LastPoint, StepPoint)
        MyBase.Step(StepPoint)
    End Sub

    Public Overrides Sub [Stop](LastPoint As Point)
        Me.Path.AddLine(Me.LastPoint, LastPoint)
        MyBase.[Stop](LastPoint)
    End Sub

    Public Overrides ReadOnly Property Bounds() As Rectangle
        Get
            Dim x, y, w, h As Integer, IsFirst As Boolean = True
            For Each item As PointF In Me.Path.PathPoints
                If IsFirst Then
                    IsFirst = False
                    x = item.X
                    y = item.Y
                    w = item.X
                    h = item.Y
                End If

                If x > item.X Then x = item.X
                If y > item.Y Then y = item.Y
                If w < item.X Then w = item.X
                If h < item.Y Then h = item.Y
            Next

            Return New Rectangle(x, y, w - x, h - y)
        End Get
    End Property

    Public Overrides Sub Paint(g As Graphics)
        If Me.Fill Then
            g.FillPath(Me.Brush, Me.Path)
        End If
        g.DrawPath(Me.Pen, Me.Path)
    End Sub
End Class

File: MorySnip/Classes/Layers/LayerFree.cs
using System.Drawing;
using System.Drawing.Drawing2D;
using Microsoft.VisualBasic.CompilerServices;

namespace Morysoft.MorySnip.Classes.Layers;

public class LayerFree : Layer
{
    public GraphicsPath Path { get; private set; } = new GraphicsPath();

    public LayerFree()
    {
    }

    public LayerFree(Pen Pen, Brush Brush, Point FirstPoint)
    {
        this.Pen = Pen;
        this.Brush = Brush;
        this.Start(FirstPoint);
    }

    public override void Step(Point StepPoint)
    {
        this.Path.AddLine(this.LastPoint, StepPoint);
        base.Step(StepPoint);
    }

    public override void Stop(Point LastPoint)
    {
        this.Path.AddLine(this.LastPoint, LastPoint);
        base.Stop(LastPoint);
    }

    public override Rectangle Bounds
    {
        get
        {
            int x = default, y = default, w = default, h = default;
            var IsFirst = true;

            foreach (var item in this.Path.PathPoints)
            {
                if (IsFirst)
                {
                    IsFirst = false;
                    x = Conversions.ToInteger(item.X);
                    y = Conversions.ToInteger(item.Y);
                    w = Conversions.ToInteger(item.X);
                    h = Conversions.ToInteger(item.Y);
                }

                if (x > item.X)
                {
                    x = Conversions.ToInteger(item.X);
                }

                if (y > item.Y)
                {
                    y = Conversions.ToInteger(item.Y);
                }

                if (w < item.X)
                {
                    w = Conversions.ToInteger(item.X);
                }

                if (h < item.Y)
                {
                    h = Conversions.ToInteger(item.Y);
                }
            }

            return new Rectangle(x, y, w - x, h - y);
        }
    }

    public override void Paint(Graphics g)
    {
        if (this.Fill)
        {
            g.FillPath(this.Brush, this.Path);
        }

        g.DrawPath(this.Pen, this.Path);
    }
}
